go/fixchain/fix_test.go

```diff
@@ -0,0 +1,244 @@
+package fixchain
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/google/certificate-transparency/go/x509"
+	"github.com/google/certificate-transparency/go/x509/pkix"
+)
+
+type fixTest struct {
+	cert string
+	chain []string
+	roots []string
+
+	function string
+	expectedChains [][]string
+	expectErr bool
+}
+
+var fixTests = []fixTest{
+	// constructChain()
+	{	// Correct chain returns chain
+		cert: googleLeaf,
+		chain: []string{verisignRoot, thawteIntermediate},
+		roots: []string{verisignRoot},
+
+		function: "constructChain",
+		expectedChains: [][]string{
+			{"Google", "Thawte", "Verisign"},
+		},
+	},
+	{	// No roots results in an error
+		cert: googleLeaf,
+		chain: []string{verisignRoot, thawteIntermediate},
+
+		function: "constructChain",
+		expectErr: true,
+	},
+	{	// No complete chain results in an error
+		cert: googleLeaf,
+		roots: []string{verisignRoot},
+
+		function: "constructChain",
+		expectErr: true,
+	},
+	{	// The wrong chain results in an error
+		cert: smimeLeaf,
+		chain: []string{verisignRoot, thawteIntermediate},
+		roots: []string{verisignRoot},
+
+		function: "constructChain",
+		expectErr: true,
+	},
+
+	// fixChain()
+	{	// Correct chain results in duplicates of chain due to behaviour of
+		// augmentIntermediates() and allowance of duplicate certs in bag.
+		// TODO(katjoyce): Would it be better to prevent duplication in bag?
```

Comment:
Having looked at the x509.CertPool.AddCert() code, it already takes care of duplicate certs within a CertPool.  Something else is going on here.